db/db_compaction_test.cc

```diff
@@ -5306,6 +5306,182 @@ TEST_F(DBCompactionTest, PersistRoundRobinCompactCursor) {
+  test::SleepingBackgroundTask sleeping_task_low;
+  env_->Schedule(&test::SleepingBackgroundTask::DoSleepTask, &sleeping_task_low,
+                 Env::Priority::LOW);
+
+  Random rnd(301);
+  // Pre-populate a 4-level tree
+  // 15 files in L2 (Flush more than 10 files may trigger auto compaction
+  // in L1 (the compact cursor may move under round-robin), so we use manual
+  // compaction twice to ensure the compact cursor is empty first.
+  for (auto i = 0; i < 9; i++) {
+    for (auto j = 0; j < 20; j++) {
+      ASSERT_OK(Put(Key(i * 60 + j * 3), rnd.RandomString(1024)));
+    }
+    ASSERT_OK(Flush());
+  }
+  MoveFilesToLevel(2);
+  for (auto i = 9; i < 15; i++) {
+    for (auto j = 0; j < 20; j++) {
+      ASSERT_OK(Put(Key(i * 60 + j * 3), rnd.RandomString(1024)));
+    }
+    ASSERT_OK(Flush());
+  }
+  MoveFilesToLevel(2);
+  // L2: [0 => 57] [60 => 117] ... [840 => 897]
+  ASSERT_EQ(15, NumTableFilesAtLevel(2));
+
+  // 9 files in L1
+  for (auto i = 0; i < 9; i++) {
+    for (auto j = 0; j < 20; j++) {
+      ASSERT_OK(Put(Key(i * 40 + j * 2 + 80), rnd.RandomString(1024)));
+    }
+    ASSERT_OK(Flush());
+  }
+  MoveFilesToLevel(1);
+  // L1: [80 => 118] [120 => 158] ... [400 => 438]
+  ASSERT_EQ(9, NumTableFilesAtLevel(1));
+
+  // 4 files in L0
+  for (auto i = 0; i < 4; i++) {
+    for (auto j = 0; j < 20; j++) {
+      ASSERT_OK(Put(Key(i * 40 + j * 2 + 440), rnd.RandomString(1024)));
+    }
+    ASSERT_OK(Flush());
+  }
+  // L0: [440 => 478] [480 => 518] [520 => 558] [560 => 598]
+  ASSERT_EQ(4, NumTableFilesAtLevel(0));
+  // Ensure max_compactions is properly applied
+  auto stop_token =
+      dbfull()->TEST_write_controler().GetCompactionPressureToken();
+
+  // Now we have a stable tree with files "4, 9, 15" in the first 3 levels.
+
+  MoveFilesToLevel(1);
+  // We will manually move 4 files from L0 to L1 and L1 will then have 13 files.
+  // When an auto compaction is triggered in L1, the first 4 files are chosen
+  // as the input for the compactions [80 => 118] [120 => 158] ... [200 => 238]
+  // In L2, 3 files [60 => 117] [120 => 177] [180 => 237] are overlapped with
+  // input files from L1. Building the boundaries involves the smallest and
+  // largestkeys from two levels and smallest keys of all files in output L2.
+  // More detailed can be found see logic in
+  // CompactionJob::GenSubcompactionBoundaries(). Therefore, we have boundaires
+  // [60, 80, 120, 180, 237, 238]. Observe that the second file in L1 has
+  // smallest key 120, the top-2 most dense range would be [120, 180] and [180,
+  // 237] (ommit [237,238]) and file [60 => 117] will be broken into two parts
+  // into [60, 80] and [80, 120]. Therefore, the average density with planned
+  // subcompaction of 3 will be more than the sum of the estimated sizes of [60,
+  // 80] and [80, 120]. This results in only 2 actual subcompactions and thus we
+  // need to release one thread at the end.
```

Comment:
A related suggestion: since the test is getting long with complicated setup, is it possible to encapsulate the setup into self-explanatory helper functions?

For reader like me (who does not know the nitty-gritty of compaction setup and is willing to trust you have set it up correctly), I probably just want to see some helper functions are there and then go straight to all your assertions related to resource reservation. After all, the test is meant for testing `RoundRobinSubcompactionsWithReservedThreads`

Reply:
> Do you mean remove these details and just keep the summary?

I just hope readers can get the main idea (the files picked are expected to form only two subcompactions, I think) without getting lost in the details like the endpoints of every file. Maybe something like:

```
  // Claim: the auto compaction for this setup will form two subcompactions
  // Proof: 
  // We will manually move 4 files from L0 to L1 and L1 will then have 13 files.
  // When an auto compaction is triggered in L1, the first 4 files are chosen
  // as the input for the compactions [80 => 118] [120 => 158] ... [200 => 238]
  // In L2, 3 files [60 => 117] [120 => 177] [180 => 237] are overlapped with
  // input files from L1. Building the boundaries involves the smallest and
  // largestkeys from two levels and smallest keys of all files in output L2.
  // More detailed can be found see logic in
  // CompactionJob::GenSubcompactionBoundaries(). Therefore, we have boundaires
  // [60, 80, 120, 180, 237, 238]. Observe that the second file in L1 has
  // smallest key 120, the top-2 most dense range would be [120, 180] and [180,
  // 237] (ommit [237,238]) and file [60 => 117] will be broken into two parts
  // into [60, 80] and [80, 120]. Therefore, the average density with planned
  // subcompaction of 3 will be more than the sum of the estimated sizes of [60,
  // 80] and [80, 120]. This results in only 2 actual subcompactions and thus we
  // need to release one thread at the end.
```

Reply:
> A related suggestion: since the test is getting long with complicated setup, is it possible to encapsulate the setup into self-explanatory helper functions?
> 
> For reader like me (who does not know the nitty-gritty of compaction setup and is willing to trust you have set it up correctly), I probably just want to see some helper functions are there and then go straight to all your assertions related to resource reservation. After all, the test is meant for testing `RoundRobinSubcompactionsWithReservedThreads`

The steps get complicated because it is a bit hard to design a case for `ShrinkSubcompactionResources`. The key range should be well controlled and we may also need a dedicated and understandable proof to explain why `num_actual_subcompactions` could be less than `num_planned_subcompactions`. We actually do not need such complicated setup for other cases. In this case, I think adding test cases without helper function would be fine?

Reply:
>  I think adding test cases without helper function would be fine?

Fine with me.